### PR TITLE
lib: Use Nix's static scope checking, fix error message, optimize

### PR DESCRIPTION
**Pull request by roberth**
##### Motivation

Nix can perform static scope checking, but whenever code is inside
a `with` expression, the analysis breaks down, because it can't
know statically what's in the attribute set whose attributes were
brought into scope. In those cases, Nix has to assume that
everything works out.

Except it doesnt. Removing `with` from lib/ revealed an undefined
variable in an error message.

If that doesn't convince you that we're better off without `with`,
I can tell you that this PR results in a 3% evaluation performance
improvement because Nix can look up local variables by index.
This adds up with applications like the module system.

Furthermore, removing `with` makes the binding site of each
variable obvious, which helps with comprehension.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
Ironically my effort to select the attributes from the right `lib.*` attrsets can't be checked by the scope checking.

Those inherits should be good now. If anyone is concerned if ofborg and me missed one, we could `inherit` from a single attrset per file like `inherit (lib // lib.trivial // ...) isFloat ...;` but that pattern smells a bit of tech debt. So I suggest using the current approach. I'll happily fix a wrong inherit if it comes up.

File: lib/debug.nix [hunk 1]
```diff
@@ -14,9 +14,25 @@
...
+    concatStringsSep
+    const
+    elem
+    generators
+    head
+    id
+    isDerivation
+    isFunction
+    mapAttrs
+    trace;
```
Comment (Profpatsch) [inline lib/debug.nix:35]: Doesn’t importing from `lib` instead of `builtins` introduce another indirection? Sounds to me like we can get another small speedup by being explicit about the builtins; otoh that means we need to be careful to not import stuff that’s not supported in old nix versions. 
Maybe worth investigating whether there’s a small evaluation gain there.

File: lib/modules.nix [hunk 2]
```diff
@@ -1,12 +1,55 @@
...
+    min
+    warn
+    ;
+  inherit (lib.strings)
+    optionalString
+    ;
+  inherit (lib)
+    elem
+    isAttrs
+    ;
```
Comment (Profpatsch) [inline lib/modules.nix]: Import list is correct (checked with the nix-mode in emacs)

File: lib/modules.nix [hunk 3]
```diff
@@ -616,7 +659,7 @@ rec {
   fixupOptionType = loc: opt:
     let
       options = opt.options or
-        (throw "Option `${showOption loc'}' has type optionSet but has no option attribute, in ${showFiles opt.declarations}.");
+        (throw "Option `${showOption loc}' has type optionSet but has no option attribute, in ${showFiles opt.declarations}.");
```
Comment (Profpatsch) [inline lib/modules.nix:660]: Is this the place that was buggy?

File: lib/options.nix [hunk 4]
```diff
@@ -1,11 +1,15 @@
...
-with lib.trivial;
-with lib.lists;
-with lib.attrsets;
-with lib.strings;
-
+let
+  inherit (lib)
+    isAttrs isBool isDerivation isFunction isInt isList isString
+    all collect concatMap concatLists elemAt filter foldl' head length mapAttrs optionals optional take
+    ;
```
Comment (Profpatsch) [inline lib/options.nix:8]: nit: the other import lists are each on their own line

File: lib/options.nix [hunk 5]
```diff
@@ -1,11 +1,15 @@
...
+  inherit (lib.attrsets) optionalAttrs;
+  inherit (lib.strings) concatMapStrings concatStringsSep;
+  inherit (lib.types) mkOptionType;
```
Comment (Profpatsch) [inline lib/options.nix:11]: correct import list

File: lib/options.nix [hunk 6]
```diff
@@ -110,7 +114,7 @@ rec {
     # Return early if we only have one element
     # This also makes it work for functions, because the foldl' below would try
     # to compare the first element with itself, which is false for functions
-    else if length defs == 1 then (elemAt defs 0).value
+    else if length defs == 1 then (head defs).value
```
Comment (Profpatsch) [inline lib/options.nix:117]: This should go into its own commit.

File: lib/sources.nix [hunk 7]
```diff
@@ -1,16 +1,33 @@
...
+    storeDir
+    tryEval
+    ;
+  inherit (lib)
+    filter
+    getAttr
+    isString
+    pathExists
+    split
+    ;
```
Comment (Profpatsch) [inline lib/sources.nix:19]: import list correct

The first file in the commit moves everything from `builtins` to `lib`, see discussion above. We should be consistent.

File: lib/strings-with-deps.nix [hunk 8]
```diff
@@ -41,10 +41,15 @@ Usage:
...
-with lib.strings;
-
+let
+  inherit (lib)
+    concatStringsSep
+    head
+    isAttrs
+    listToAttrs
+    tail
+    ;
```
Comment (Profpatsch) [inline lib/strings-with-deps.nix:51]: import list correct

File: lib/strings.nix [hunk 9]
```diff
@@ -8,7 +8,29 @@ in
...
+    replaceStrings
+    split
+    storeDir
+    stringLength
+    substring
+    tail
+    toJSON
+    typeOf
+    unsafeDiscardStringContext
+    ;
```
Comment (Profpatsch) [inline lib/strings.nix:33]: import list correct

File: lib/types.nix [hunk 10]
```diff
@@ -1,12 +1,62 @@
...
+    mergeOneOption
+    showFiles
+    showOption
+    ;
+  inherit (lib.strings)
+    concatMapStringsSep
+    concatStringsSep
+    escapeNixString
+    isCoercibleToString
+    ;
```
Comment (Profpatsch) [inline lib/types.nix:59]: import lists correct

File: lib/types.nix [hunk 11]
```diff
@@ -270,7 +320,7 @@ rec {
       name = "attrs";
       description = "attribute set";
       check = isAttrs;
-      merge = loc: foldl' (res: def: mergeAttrs res def.value) {};
+      merge = loc: foldl' (res: def: res // def.value) {};
```
Comment (Profpatsch) [inline lib/types.nix:323]: should go into its own commit

File: lib/default.nix [hunk 12]
```diff
@@ -65,7 +65,7 @@ let
       lessThan listToAttrs pathExists readFile replaceStrings seq
       stringLength sub substring tail;
     inherit (self.trivial) id const pipe concat or and bitAnd bitOr bitXor
-      bitNot boolToString mergeAttrs flip mapNullable inNixShell min max
+      bitNot boolToString mergeAttrs flip mapNullable inNixShell isFloat min max
```
Comment (Profpatsch) [inline lib/default.nix:68]: how was it exposed before?

File: lib/default.nix [hunk 13]
```diff
@@ -63,7 +63,7 @@ let
       deepSeq elem elemAt filter genericClosure genList getAttr
       hasAttr head isAttrs isBool isInt isList isString length
       lessThan listToAttrs pathExists readFile replaceStrings seq
-      stringLength sub substring tail;
+      stringLength sub substring tail trace;
```
Comment (Profpatsch) [inline lib/default.nix:66]: This has been bothering me for a while :+1:

**Review by Profpatsch — CHANGES_REQUESTED** [review]
Great success!

Ofborg is green without any rebuilds, and I checked the import lists with the static mode, so it looks like this is a no-op (apart from the new symbols exposed).

**Comment by Profpatsch** [comment]
Some of my comments only make sense in the full diff view, Github is not a good tool for these reviews.

File: lib/debug.nix (hunk 1 above)
Comment (roberth) [reply lib/debug.nix:35]: I had the same thought, but `lib`s top-level attribute set is only evaluated once, so the overhead of this indirection amounts to a single lookup. You really need the repeated evaluation of certain expressions to make this optimization effective.

I'd be in favor of adding more builtins to `lib` for consistency and convenience. I've only done it for ones I got wrong writing the explicit inherits, which, although admittedly a bit arbitrarily, keeps the scope of this PR in check.

File: lib/modules.nix (hunk 3 above)
Comment (roberth) [reply lib/modules.nix:660]: Exactly. Scope checking will now find undeclared local variables, even if not used in any tests.

File: lib/default.nix (hunk 12 above)
Comment (roberth) [reply lib/default.nix:68]: `lib.trivial` only, unlike the other three `lib.trivial.is*`

**Comment by roberth** [comment]
Rebased and recent lib additions tested.

File: lib/options.nix (hunk 4 above)
Comment (roberth) [reply lib/options.nix]: Done. Using merge-friendly syntax now.

File: lib/options.nix (hunk 6 above)
Comment (roberth) [reply lib/options.nix:142]: Done.

File: lib/types.nix (hunk 11 above)
Comment (roberth) [reply lib/types.nix:323]: Done.

File: lib/sources.nix (hunk 7 above)
Comment (roberth) [reply lib/sources.nix]: This only applies to `readFile`, which I've just done. I'd like to leave adding more builtins to lib out of scope for this PR.

**Comment by roberth** [comment]
I appreciate further improvements to `lib` but I'd like to merge this soon to avoid conflicts with other changes to lib.

File: lib/debug.nix (hunk 1 above)
Comment (Profpatsch) [reply lib/debug.nix:35]: Let’s do it in a followup then.

File: lib/types.nix [hunk 14]
```diff
@@ -270,7 +270,7 @@ rec {
...
```
Comment (Profpatsch) [inline lib/types.nix:323]: `mergeAttrs = x: y: x // y` so this is correct

**Review by Profpatsch — APPROVED** [review]
Looks good to me!

**Comment by roberth** [comment]
I've waited a bit because I wasn't going to be available in the past couple of days. No relevant changes were made to `lib/` in the meantime, so I'll go ahead with the merge.

**Comment by vbgl** [comment]
Evaluation seems broken:

> error: undefined variable 'boolToString' at …/nixpkgs/lib/types.nix:552:42

**Comment by xaverdh** [comment]
This is coming from https://github.com/NixOS/nixpkgs/commit/cebf9198f3fe80fa38a8c57eb3b02ac4c1df3e8a#diff-64168148acd9f2147ef733b1498b8821c24f2e3f32354b0e147dd421d71274f3, which was merged in the meantime

**Comment by roberth** [comment]
Apologies for the breakage.
I did check the diff but I must have misread it.

**Comment by deviant** [comment]
FWIW, this [breaks](https://github.com/NixOS/nixpkgs/pull/102191) `lib.commitIdFromGitRepo`.

**Comment by Profpatsch** [comment]
Was fixed.
